libirods_rule_engine_plugin-indexing.cpp

```diff
@@ -697,6 +740,151 @@ namespace {
+        std::map<std::string,int> user_entry;
+        std::multimap<int,std::string> user_perms, group_perms;
+
+        static std::map<int,std::string> perm_names;
+
+        bool is_group( const std::string &gid ) {
+            try {
+                groups.at(gid);
+            }
+            catch (const std::out_of_range&) {
+                return false;
+            }
+            return true;
+        }
+
+        bool is_member_of (const std::string& user_id, const std::string& group_id) {
+            try {
+                const auto & user_list = members.at(group_id);
+                return std::find( user_list.begin(), user_list.end(), user_id) != user_list.end();
+            }
+            catch (...) {
+              std::cerr << ("unknown error\n");
+              throw;
+            }
+            return false;
+        }
+
+        void calc_perm_info( const std::string & obj_id, const std::string & obj_type);
+
+        rsComm_t *Conn {};
+        std::string owner {};
+
+        void calc_user_info_() {
+            irods::query q {Conn, "select USER_GROUP_NAME,USER_GROUP_ID,USER_NAME,USER_ID"};
+            for (const auto & c:q) {
+                if (c[1] != c[3]) {
+                    members[c[1]].push_back(c[3]);
+                    groups[c[1]]=c[0];
+                }
+                else {
+                    users[c[3]]=c[2];
+                }
+            }
+        }
+
+        int idx {0};
+
+        void reset_perms_ () {
+            if (idx != 0) {
+                user_entry = {};
+                group_perms = {};
+                user_perms = {};
+                owner = {};
+            }
+            idx = 0;
+        }
+
+    public:
+
+        void get_perms_list(   nlohmann::json & j,
+                               const std::string & obj_id, const std::string & obj_type )
+        {
+            calc_perm_info( obj_id, obj_type) ;
+
+            for (const auto & [pm,gid] : group_perms) {
+                j["userPermissions"][idx]["permission"] = perm_names.at(pm);
+                j["userPermissions"][idx++]["user"] = groups[gid];
+            }
+            for (const auto & [pm,uid] : user_perms) {
+                j["userPermissions"][idx]["permission"] = perm_names.at(pm);
+                j["userPermissions"][idx++]["user"] = users[uid];
+            }
+            j["creator"] = owner;
+            idx = -1;  // force structures to be reinitialized on reuse
+        }
+
+        PermsCalc(rsComm_t *Conn_): Conn{Conn_}
+        {
+            calc_user_info_();
+        }
+
+        PermsCalc(const PermsCalc & x, rsComm_t *Conn_) : users{x.users}, groups{x.groups}, members{x.members}, Conn{Conn_}
+        {
+            reset_perms_();
+        }
+
+        static std::unique_ptr<PermsCalc> ptr ; // lazy instantiation
+    };
+
+    std::map<int,std::string> PermsCalc::perm_names {
+        { 1050 , "read" },
+        { 1120 , "write" },
+        { 1200 , "own"}
+    };
```

Comment:
catalog_utilities.hpp (access_type).
store enumeration  as its underlying type

Reply:
1. `catalog_utilities.hpp` not in irods-dev 
2. not trivial and may break algorithm. Maybe skip this one @trel @korydraughn  ?

Reply:
You're right. Skip it, but create an issue for later.

Reply:
issue was created.